### PR TITLE
add pod antiaffinity.

### PR DESCRIPTION
**Pull request by morvencao**
PR based on https://github.com/istio/istio/pull/12464

fixes: #11333

### CODE REVIEW

File: install/kubernetes/helm/istio/charts/grafana/values.yaml [hunk 1]
```diff
@@ -28,6 +28,28 @@ security:
...
+#    "requiredDuringSchedulingIgnoredDuringExecution"
+#    "preferredDuringSchedulingIgnoredDuringExecution"
+# which denote “hard” vs. “soft” requirements, you can define your values
+# in "podAntiAffinityLabelSelector" and "podAntiAffinityTermLabelSelector"
+# correspondingly.
+# For example:
+# podAntiAffinityLabelSelector:
+# - key: security
+#   operator: In
+#   value: S1
```
Comment (linsun) [inline install/kubernetes/helm/istio/charts/grafana/values.yaml]: should this be values?

Comment (linsun) [reply install/kubernetes/helm/istio/charts/grafana/values.yaml]: also does the key or operator matters to our users?  any reason why user needs to configure this?

File: install/kubernetes/helm/istio/charts/galley/values.yaml [hunk 2]
```diff
@@ -6,3 +6,23 @@ replicaCount: 1
...
+#    "requiredDuringSchedulingIgnoredDuringExecution"
+#    "preferredDuringSchedulingIgnoredDuringExecution"
+# which denote “hard” vs. “soft” requirements, you can define your values
+# in "podAntiAffinityLabelSelector" and "podAntiAffinityTermLabelSelector"
+# correspondingly.
+# For example:
+# podAntiAffinityLabelSelector:
+# - key: security
+#   operator: In
+#   value: S1
```
Comment (linsun) [inline install/kubernetes/helm/istio/charts/galley/values.yaml]: would be good to give examples of multiple values.

File: install/kubernetes/helm/istio/charts/gateways/templates/_affinity.tpl [hunk 3]
```diff
@@ -41,3 +41,55 @@
...
+{{- end }}
+
+{{- define "gatewaypodAntiAffinityPreferredDuringScheduling" }}
+    {{- range $index, $item := .podAntiAffinityTermLabelSelector }}
+    - podAffinityTerm:
+        labelSelector:
+          matchExpressions:
+          - key: {{ $item.key }}
+            operator: {{ $item.operator }}
+            {{- if .value }}
```
Comment (linsun) [inline install/kubernetes/helm/istio/charts/gateways/templates/_affinity.tpl]: is there any value to list key or operator if there is no .values?

Comment (morvencao) [reply install/kubernetes/helm/istio/charts/gateways/templates/_affinity.tpl]: If the operator is `Exists` or `DoesNotExist`, there is no `value`

File: install/kubernetes/helm/istio/charts/gateways/templates/_affinity.tpl [hunk 4]
```diff
@@ -41,3 +41,55 @@
...
+      topologyKey: {{ $item.topologyKey }}
+    {{- end }}
+{{- end }}
+
+{{- define "gatewaypodAntiAffinityPreferredDuringScheduling" }}
+    {{- range $index, $item := .podAntiAffinityTermLabelSelector }}
+    - podAffinityTerm:
+        labelSelector:
+          matchExpressions:
+          - key: {{ $item.key }}
```
Comment (linsun) [inline install/kubernetes/helm/istio/charts/gateways/templates/_affinity.tpl]: wonder if we should check earlier for existence of key, operator and values and topologyKey.... or should we even default some of these when not specified?   like key: security operator: In.

File: install/kubernetes/helm/istio/templates/_affinity.tpl [hunk 5]
```diff
@@ -40,4 +40,56 @@
...
+            {{- if .value }}
+            values:
+            {{- $vals := split "," $item.value }}
+            {{- range $i, $v := $vals }}
+            - {{ $v }}
+            {{- end }}
+            {{- end }}
+        topologyKey: {{ $item.topologyKey }}
+      {{- if $item.weight }}
+      weight: {{ $item.weight }}
```
Comment (linsun) [inline install/kubernetes/helm/istio/templates/_affinity.tpl]: should we have a default weight 100?  any reason why user needs to configure weight other than 100?

File: install/kubernetes/helm/istio/charts/galley/values.yaml (hunk 2 above)
Comment (morvencao) [reply install/kubernetes/helm/istio/charts/galley/values.yaml]: Yes, multiple `values` is also support, separated by comma, here is my local test file: https://github.com/istio/istio/pull/12691/commits/95748adce258b7abc39172357a26b7ef6ef7e0ef#diff-1aa45219f5b1d3d1013e6a11ce46ce37

File: install/kubernetes/helm/istio/templates/_affinity.tpl (hunk 5 above)
Comment (morvencao) [reply install/kubernetes/helm/istio/templates/_affinity.tpl]: > The weight field in preferredDuringSchedulingIgnoredDuringExecution is in the range 1-100.

Seems that we can specify `weight` to multiple podAffinityTerm

File: install/kubernetes/helm/istio/charts/gateways/templates/_affinity.tpl (hunk 4 above)
Comment (morvencao) [reply install/kubernetes/helm/istio/charts/gateways/templates/_affinity.tpl]: We can add more check for these values, but default value should not make sense, IMO.

File: install/kubernetes/helm/istio/charts/gateways/templates/_affinity.tpl (hunk 3 above)
Comment (linsun) [reply install/kubernetes/helm/istio/charts/gateways/templates/_affinity.tpl]: ok, fair point!

File: install/kubernetes/helm/istio/templates/_affinity.tpl (hunk 5 above)
Comment (linsun) [reply install/kubernetes/helm/istio/templates/_affinity.tpl]: yes, i saw your example that you can configure weight 50 for one labelselector, and another 50 for another labelselector.  I understand the k8s pod anti-affnity platform can support all these config nobs, but the question is if we need to expose everyone of them to our users?  If we don't have a common scenario to use them on istio components, I'd rather not expose them.

Comment (morvencao) [reply install/kubernetes/helm/istio/templates/_affinity.tpl]: OK, got it.

**Pull request merged**